features/22-email-verification/implementation.md

```diff
@@ -0,0 +1,216 @@
+# Implementation for verify by email and SMS
+
+## Server
+
+The server (core) will implement the following set of APIs to support
+verify by email and verify by SMS
+
+### `auth:verify:check`
+
+#### Overview
+
+#### Request
+
+* `verify_key` (string)
+  The auth key, either `email` or `phone`.
+
+* `user_id` (string, optional)
+  The user to verify. If the request is authenticated with access token
+  only the current user is accepted. If request is authenticated with master key
+  any user is accepted. Default to current user. If the request is not
+  authenticated, the user will be looked up from `verify_data`.
+
+* `verify_data` (string, optional)
+  The auth key value, should be email if `verify_key` is `email` or phone
+  if `verify_key` is `phone`.
+  If not specified, the current email or phone is used.
+
+* `code` (string)
+  The verification code.
+
+#### Response
+
+If verification is a success, a `authResponse` is returned.
+
+Example:
+
+```
+{
+    "user_id": "E7BEE2DE-445B-42D7-B90A-84CB735A25D6",
+    "profile": {
+        "email": "johndoe@example.com"
+    },
+    "access_token": "eyJhbGciOiJIUzI1Ni.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6Ikp.TJVA95OrM7E2c",
+    "last_login_at": "2017-07-23T19:30:24Z",
+    "verified": true,
+    "verified_record_keys": {
+        "email": true
+    }
+}
+```
+
+If code is invalid, `InvalidArgument` error is returned.
+
+### `auth:verify:send`
+
+#### Overview
+
+#### Request
+
+* `verify_key` (string)
+  The auth key, either `email` or `phone`.
+
+* `user_id` (string, optional)
+  The user to verify. If the request is authenticated with access token
+  only the current user is accepted. If request is authenticated with master key
+  any user is accepted. Default to current user. If the request is not
+  authenticated, the user will be looked up from `verify_data`.
+
+* `verify_data` (string, optional)
+  The auth key value, should be email if `verify_key` is `email` or phone
+  if `verify_key` is `phone`.
+  If not specified, the current email or phone is used.
+
+#### Response
+
+The API returns Status OK when sent successfully.
+
+### `auth:*` that returns an authResponse
+
+For all API that returns an authResponse, the API has to be modified
+to include the following fields:
+
+* `verified` (boolean)
+  If the user is verified with one of the keys, this returns true.
+
+* `verified_record_keys` (object, string: any)
+  Returns which piece of auth record key is verified.
+
+APIs that include authResponse are (but not limited to):
+
+* `auth:login`
+* `auth:signup`
```

Comment:
Also the sso api
`sso:oauth:login`
`sso:oauth:signup`
`sso:custom_token:login`

Reply:
thanks

Reply:
and can I specify type of verification in server request? Or there is only one type of verification in one skygear server?

Reply:
and do we have role ACL for verified user and unverified user?

Reply:
and in (python cloud function )[https://docs.skygear.io/guides/cloud-function/lambda/python/],
there is `@skygear.op(name, user_required=False, key_required=False)`
do we need to add `user_verified=False/True`?

Reply:
@howawong 
> non admin user calls verify_user and unverify_user
The response will be the same as the other apis that require master key and admin, it should be deny

> an I specify type of verification in server request
The type of verification is defined by environment variable, one skygear server can have multiple verification method (email, phone...etc). But one user field can only define one verification method. You can see the detail in configuration section.

> do we have role ACL for verified user and unverified user
We don't have this plan yet, but we can plan for this if needed in future

> do we need to add user_verified=False/True
To avoid confusion, we will not add this now. But we will have verified flag in the context, so developer can achieve it by checking the context